tinylog-impl/src/main/java/org/tinylog/writers/JsonWriter.java

```diff
@@ -0,0 +1,268 @@
+		if (hasLevel) {
+			writeProperty(levelToken, logEntry, builder, hasTimestamp);
+		}
+		if (hasClass) {
+			writeProperty(classToken, logEntry, builder, hasLevel);
+		}
+		if (hasMethod) {
+			writeProperty(methodToken, logEntry, builder, hasClass);
+		}
+		if (hasThread) {
+			writeProperty(threadToken, logEntry, builder, hasMethod);
+		}
+		return String.format(JSON_OBJECT, builder.toString());
+	}
+
+	@Override
+	public void write(final LogEntry logEntry) throws IOException {
+		String jsonLogEntry;
+		if (builder == null) {
+			builder = new StringBuilder();
+		} else {
+			builder.setLength(0);
+		}
+		jsonLogEntry = buildJsonLogEntry(logEntry, builder);
+		jsonLogEntry += ",";
+		writer.write(jsonLogEntry.getBytes(charset), jsonLogEntry.length());
+	}
+
+	@Override
+	public void flush() throws IOException {
+		writer.flush();
+	}
+
+	@Override
+	public void close() throws IOException {
+		postprocessFile();
+		writer.close();
+	}
+
+	/**
+	 * Preprocesses the JSON file. If append mode is on, deletes the closing bracket
+	 * and adds a comma instead. If it's a new file, appends an opening bracket.
+	 * 
+	 * @param append Append Mode on or off
+	 * @throws IOException              Error reading or writing file
+	 * @throws IllegalArgumentException Invalid file format
+	 */
+	private void preprocessFile(final boolean append) throws IOException, IllegalArgumentException {
+		if (append && inputChannel.size() > 0) {
+			long sizeToTruncate = 0;
+			long currentPosition = inputChannel.size();
+			boolean foundClosingBracket = false;
+			while (!foundClosingBracket) {
+				long from = Math.max(0, currentPosition - BUFFER_SIZE);
+				long numberOfBytes = Math.min(currentPosition, BUFFER_SIZE);
+				MappedByteBuffer section = inputChannel.map(FileChannel.MapMode.READ_ONLY, from, numberOfBytes);
+				byte[] bytes = new byte[section.remaining()];
+				section.get(bytes);
+				if (bytes.length == 0) {
+					throw new IllegalArgumentException(
+							"Invalid JSON file. The file is missing a closing bracket for the array.");
+				}
+				for (int i = bytes.length - 1; i >= 0; i--, currentPosition--) {
+					sizeToTruncate += 1;
+					if (bytes[i] == ']') {
+						foundClosingBracket = true;
+						break;
+					}
+				}
+			}
+			long newFileSize = fileChannel.size() - sizeToTruncate;
+			fileChannel.truncate(newFileSize);
+			writer.write(commaByte, 1);
+		}
+
+		if (inputChannel.size() == 0) {
+			writer.write(bracketOpenByte, 1);
+		}
+	}
+
+	/**
+	 * Postprocesses the JSON file. Attempts to delete the trailing comma and
+	 * appends closing bracket which were handled in
+	 * {@link #preprocessFile(boolean)}
+	 * 
+	 * @throws IOException Error writing to file
+	 */
+	private void postprocessFile() throws IOException {
+		fileChannel.truncate(fileChannel.size() - commaByte.length);
+		writer.write(bracketCloseByte, 1);
+	}
+
+	@Override
+	public Collection<LogEntryValue> getRequiredLogEntryValues() {
+		return EnumSet.noneOf(LogEntryValue.class);
```

Comment:
Here, tinylog expects all required log entry values from all used tokens, created via the format pattern parsers. You can just create a new enum set, iterate over all tokens, and add their required log entry values to the enum set. If you return, an empty enum set, it is not guaranteed that further tinylog version won't pass only empty log entries to the JSON parser as optimization result.

Reply:
Fixed in https://github.com/tinylog-org/tinylog/pull/195/commits/5322e7bbaaa6040dfad72f601cac39c19db08998